File: frontend/src/services/translation-service.ts
import { apiService } from './api'

export interface TranslationRequest {
    source_text: string
    source_language_code: string
    target_language_codes: string[]
}

export interface TranslationResponse {
    translated_text: string
    detected_language?: string
    confidence?: number
    error?: string
}

export interface MultilingualTranslationResponse {
    source_text: string
    source_language: string
    target_languages: string[]
    translations: {
        [key: string]: TranslationResponse
    }
    status: string
    error?: string
}

export interface TranslationProgress {
    step: 'starting' | 'translating' | 'completed' | 'error'
    message: string
    progress: number
    error?: string
}

class TranslationService {
    private baseEndpoint = '/translation'

    /**
     * Translate text to multiple languages
     */
    async translateText(
        sourceText: string,
        sourceLanguage: string = 'en',
        targetLanguages: string[] = ['hi', 'mr', 'gu'],
        onProgress?: (progress: TranslationProgress) => void
    ): Promise<{
        success: boolean
        translations?: {
            [key: string]: string
        }
        error?: string
    }> {
        try {
            // Initialize progress
            onProgress?.({
                step: 'starting',
                message: 'Initializing translation...',
                progress: 10
            })

            // Update progress - preparing request
            onProgress?.({
                step: 'translating',
                message: 'Sending translation request...',
                progress: 30
            })

            const response = await apiService.request<MultilingualTranslationResponse>(
                `${this.baseEndpoint}/translate`,
                {
                    method: 'POST',
                    body: JSON.stringify({
                        source_text: sourceText,
                        source_language_code: sourceLanguage,
                        target_language_codes: targetLanguages
                    })
                }
            )

            if (!response) {
                throw new Error('No response received from translation service')
            }

            // Update progress - processing response
            onProgress?.({
                step: 'translating',
                message: 'Processing translations...',
                progress: 70
            })

            // Extract translations
            const translations: { [key: string]: string } = {}
            for (const lang of targetLanguages) {
                translations[lang] = response.translations[lang]?.translated_text || ''
            }

            // Check if we got valid translations
            const hasTranslations = Object.values(translations).some(text => text.trim() !== '')

            if (!hasTranslations) {
                throw new Error('No translations received from the server')
            }

            // Update progress - completed
            onProgress?.({
                step: 'completed',
                message: 'Translation completed successfully!',
                progress: 100
            })

            return {
                success: true,
                translations
            }

        } catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Unknown error occurred'

            onProgress?.({
                step: 'error',
                message: 'Translation failed',
                progress: 0,
                error: errorMessage
            })

            return {
                success: false,
                error: errorMessage
            }
        }
    }

    /**
     * Get supported languages
     */
    async getSupportedLanguages(): Promise<{
        success: boolean
        languages?: Array<{
            language_code: string
            display_name: string
            support_source: boolean
            support_target: boolean
        }>
        error?: string
    }> {
        try {
            const response = await apiService.request<{
                languages: Array<{
                    language_code: string
                    display_name: string
                    support_source: boolean
                    support_target: boolean
                }>
                total_count: number
            }>(`${this.baseEndpoint}/supported-languages`)

            if (!response) {
                throw new Error('No response received from translation service')
            }

            return {
                success: true,
                languages: response.languages
            }

        } catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Unknown error occurred'
            return {
                success: false,
                error: errorMessage
            }
        }
    }

    /**
     * Health check for translation service
     */
    async healthCheck(): Promise<{
        success: boolean
        status?: string
        message?: string
        error?: string
    }> {
        try {
            const response = await apiService.request<{
                status: string
                message: string
                supported_languages_count: number
            }>(`${this.baseEndpoint}/health`)

            if (!response) {
                throw new Error('No response received from translation service')
            }

            return {
                success: true,
                status: response.status,
                message: response.message
            }

        } catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Unknown error occurred'
            return {
                success: false,
                error: errorMessage
            }
        }
    }
}

// Export singleton instance
export const translationService = new TranslationService()

File: backend/app/services/translation_service.py
"""
GCP Cloud Translation Service
Handles multilingual text translation using Google Cloud Translation API
"""

import os
from typing import List, Dict, Any
from pathlib import Path
from google.cloud import translate_v3
from google.oauth2 import service_account
from app.utils.logger import get_logger

logger = get_logger(__name__)


class TranslationService:
    def __init__(self):
        self.project_id = "aipower-467603"  # From the credentials file
        # Get the absolute path to the credentials file
        current_dir = Path(__file__).parent
        project_root = current_dir.parent.parent.parent  # Go up to project root
        self.credentials_path = project_root / "backend" / \
            "credentials" / "secrets" / "isl.json"
        self.client = None
        self._initialize_client()

    def _initialize_client(self):
        """Initialize the Google Cloud Translation client"""
        try:
            # Convert Path object to string for environment variable
            credentials_path_str = str(self.credentials_path)

            # Set the credentials file path
            os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = credentials_path_str

            # Create credentials from service account file
            credentials = service_account.Credentials.from_service_account_file(
                credentials_path_str,
                scopes=["https://www.googleapis.com/auth/cloud-translation"]
            )

            # Initialize the client
            self.client = translate_v3.TranslationServiceClient(
                credentials=credentials)
            logger.info(
                "Google Cloud Translation client initialized successfully")

        except Exception as e:
            logger.error(
                f"Failed to initialize Google Cloud Translation client: {str(e)}")
            raise

    def translate_text(
        self,
        text: str,
        source_language_code: str,
        target_language_codes: List[str]
    ) -> Dict[str, Any]:
        """
        Translate text from source language to multiple target languages

        Args:
            text: The text to translate
            source_language_code: Source language code (e.g., 'en', 'hi', 'mr', 'gu')
            target_language_codes: List of target language codes (e.g., ['hi', 'mr', 'gu'])

        Returns:
            Dictionary containing translations for each target language
        """
        try:
            if not self.client:
                raise Exception("Translation client not initialized")

            parent = f"projects/{self.project_id}/locations/global"
            mime_type = "text/plain"

            # Prepare the response structure
            translations = {}

            # Translate to each target language
            for target_lang in target_language_codes:
                try:
                    response = self.client.translate_text(
                        contents=[text],
                        parent=parent,
                        mime_type=mime_type,
                        source_language_code=source_language_code,
                        target_language_code=target_lang,
                    )

                    # Extract the translated text
                    if response.translations:
                        translated_text = response.translations[0].translated_text
                        translations[target_lang] = {
                            "translated_text": translated_text,
                            "detected_language": response.translations[0].detected_language_code,
                            "confidence": getattr(response.translations[0], 'confidence', None)
                        }
                    else:
                        translations[target_lang] = {
                            "translated_text": text,  # Fallback to original text
                            "error": "No translation returned"
                        }

                except Exception as e:
                    logger.error(
                        f"Error translating to {target_lang}: {str(e)}")
                    translations[target_lang] = {
                        "translated_text": text,  # Fallback to original text
                        "error": f"Translation failed: {str(e)}"
                    }

            return {
                "source_text": text,
                "source_language": source_language_code,
                "target_languages": target_language_codes,
                "translations": translations,
                "status": "success"
            }

        except Exception as e:
            logger.error(f"Translation service error: {str(e)}")
            return {
                "source_text": text,
                "source_language": source_language_code,
                "target_languages": target_language_codes,
                "translations": {},
                "status": "error",
                "error": str(e)
            }

    def get_supported_languages(self) -> List[Dict[str, str]]:
        """
        Get list of supported languages

        Returns:
            List of dictionaries containing language codes and names
        """
        try:
            if not self.client:
                raise Exception("Translation client not initialized")

            parent = f"projects/{self.project_id}/locations/global"

            response = self.client.get_supported_languages(parent=parent)

            languages = []
            for language in response.languages:
                languages.append({
                    "language_code": language.language_code,
                    "display_name": language.display_name,
                    "support_source": language.support_source,
                    "support_target": language.support_target
                })

            return languages

        except Exception as e:
            logger.error(f"Error getting supported languages: {str(e)}")
            return []


# Create a singleton instance
translation_service = TranslationService()
